### PR TITLE
New feature flags backend (part 2)

### PR DESCRIPTION
**Pull request by fisx**
## migration plan

- merge #813.
  - this will create the new team_features table, but not use it for legalhold.
  - all teams with sso configured will continue working normally, but report as "sso disabled".  (weird, yes.  sorry.)
- migrate data:
  - for every team that has an idp, set the "sso enabled" flag in galley, in the new table.
  - we'll let legalhold break on all existing teams that have it enabled (not newly created ones).
- merge #817
  - switch legalhold feature flag lookup and update from old to new table.
  - this also breaks legalhold.
- yet another PR:
  - this will remove the old legalhold feature flag table.

## tasks

- [x] make "put /i/sso-disabled" fail with an explanation why implementing this is hard.
- [ ] write (and run) migration script for SSO flag.  (frontend depends on this to be sound.)

### CODE REVIEW

File: services/galley/schema/src/V35.hs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+module V35 (migration) where
+
+import Imports
+import Cassandra.Schema
+import Text.RawString.QQ
+
+migration :: Migration
```
Comment (tiago-loureiro) [inline services/galley/schema/src/V35.hs]:
```suggestion
migration :: Migration
migration = Migration 35 "Delete deprecated legalhold_team_config"
```

File: services/galley/schema/src/V35.hs [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+migration = Migration 34 "Delete deprecated legalhold_team_config" $ do
```
Comment (tiago-loureiro) [inline services/galley/schema/src/V35.hs]: I'd suggest _not_ adding this migration on this PR; the main concern I have with this is: we will follow the playbook and run the migration before deploying new services and that will break the currently running service, which is why I left the comment.

I don't think it harms to make a separate PR and merge it after but I leave the decision to the PR maker

File: services/galley/src/Galley/API/Error.hs [hunk 3]
```diff
@@ -135,3 +135,16 @@ userLegalHoldNotPending = Error status412 "legalhold-not-pending" "legal hold ca
...
+disableSsoNotImplemented = Error status403 "not-implemented"
+    "The SSO feature flag is disabled by default.  It can only be enabled once for any team, never disabled.\n\
+    \\n\
+    \Rationale: there are two services in the backend that need to keep their status in sync: galley (teams),\n\
+    \and spar (SSO).  Galley keeps track of team features.  If galley creates an idp, the feature flag is\n\
+    \checked.  For authentication, spar avoids this expensive check and assumes that the idp can only have\n\
+    \been created if the SSO is enabled.  This assumption does not hold any more if the switch is turned off\n\
+    \again, so we do not support this.\n\
+    \\n\
+    \It is definitely feasible to change this.  Tf you have a use case, please contact customer support, or\n\
```
Comment (tiago-loureiro) [inline services/galley/src/Galley/API/Error.hs]: `Tf -> If`

File: services/galley/src/Galley/Data/SSO.hs [hunk 4]
```diff
@@ -16,7 +16,8 @@ getSSOTeamConfig :: MonadClient m => TeamId -> m (Maybe SSOTeamConfig)
 getSSOTeamConfig tid = fmap toLegalHoldTeamConfig <$> do
```
Comment (tiago-loureiro) [inline services/galley/src/Galley/Data/SSO.hs]:
```suggestion
getSSOTeamConfig tid = fmap toSSOTeamConfig <$> do
```

File: tools/db/migrate-sso-feature-flag/README.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+## Service backfill
+
+A tool for filling tables `service_user` and `service_team` from existing data.
```
Comment (tiago-loureiro) [inline tools/db/migrate-sso-feature-flag/README.md]: Please check this comment

File: tools/db/migrate-sso-feature-flag/src/Work.hs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+import qualified Data.Conduit.List as C
+import qualified System.Logger as Log
+
+deriving instance Cql Name
+
+runCommand :: Logger -> ClientState -> ClientState -> IO ()
+runCommand l spar galley = do
+    runConduit
+        $ zipSources (C.sourceList [(1::Int32) ..])
```
Comment (tiago-loureiro) [inline tools/db/migrate-sso-feature-flag/src/Work.hs]: Fancy! :)

Comment (fisx) [reply tools/db/migrate-sso-feature-flag/src/Work.hs]: We just stole it from @neongreen  :-)

Comment (neongreen) [reply tools/db/migrate-sso-feature-flag/src/Work.hs]: I can't even recall writing it 🤔At any rate, you're welcome!

**Comment by fisx** [comment]
Migration works.  I will push another commit soon that irons out a few wrinkles.  Here is the log from staging:

```
spar> select team, idp from spar.idp where team=cf0e5a44-257a-4640-ae3a-b0e9e4fe24e3 allow filtering;

 team                                 | idp
--------------------------------------+--------------------------------------
 cf0e5a44-257a-4640-ae3a-b0e9e4fe24e3 | e64ed7f7-8814-4155-ac83-d3b07f947401
```

```
galley> select * from galley.team_features where team_id=cf0e5a44-257a-4640-ae3a-b0e9e4fe24e3 ;

 team_id                              | legalhold_status | sso_status
--------------------------------------+------------------+------------
 cf0e5a44-257a-4640-ae3a-b0e9e4fe24e3 |             null |          0
```

```
export GALLEY_HOST=...
export GALLEY_KEYSPACE=...
export SPAR_HOST=...
export SPAR_KEYSPACE=...

ssh -v -f ubuntu@${GALLEY_HOST} -L 2021:${GALLEY_HOST}:9042 -N
ssh -v -f ubuntu@${SPAR_HOST} -L 2022:${SPAR_HOST}:9042 -N

./dist/migrate-sso-feature-flag --cassandra-host-galley=localhost --cassandra-port-galley=2021 --cassandra-keyspace-galley=${GALLEY_KEYSPACE} --cassandra-host-spar=localhost --cassandra-port-spar=2022 --cassandra-keyspace-spar=${SPAR_KEYSPACE}
D, Connecting to ::1:2022
I, New control connection: eu-west:1a:::1:2022#12
D, Connecting to ::1:2021
::1 port 41206 to ::1 port 2021, nchannels 4
D, Connection established: eu-west:1a:::1:2022#14
D, Connection established: eu-west:1a:::1:2022#17
I, convs=10
D, Connection established: eu-west:1a:::1:2021#15
I, convs=20
I, convs=30
I, convs=40
I, convs=50
[...]
I, convs=3570
I, convs=3580
I, convs=3590
I, convs=3600
I, convs=3610
I, convs=3620
I, convs=3630
I, convs=3640
I, convs=3650
I, convs=3660
I, convs=3670
I, convs=3680
I, convs=3690
I, convs=3700
I, convs=3710
I, convs=3720

$ echo $?
0
```

```
galley> select * from galley.team_features where team_id=cf0e5a44-257a-4640-ae3a-b0e9e4fe24e3 ;

 team_id                              | legalhold_status | sso_status
--------------------------------------+------------------+------------
 cf0e5a44-257a-4640-ae3a-b0e9e4fe24e3 |             null |          1
```

File: services/galley/schema/src/V35.hs (hunk 2 above)
Comment (fisx) [reply services/galley/schema/src/V35.hs]: summary of off-line discussion: generally yes, this should be spread out over two PRs, but since we promised to break staging and there is no legalhold code in prod yet, we're going to make an exception this time, and do it in one step.
